lib/formatters/sarif.js

```diff
@@ -0,0 +1,182 @@
+ */
+class SarifPrinter {
+  constructor(options = {}) {
+    this.options = options;
+    this.console = options.console || console;
+  }
+
+  print(results) {
+    const sarifLog = {
+      version: '2.1.0',
+      $schema: 'http://json.schemastore.org/sarif-2.1.0-rtm.5',
+      runs: [
+        {
+          tool: {
+            driver: {
+              name: 'ember-template-lint',
+              informationUri: 'https://github.com/ember-template-lint/ember-template-lint',
+              rules: [],
+              version: require('../../package.json').version,
+            },
+          },
+        },
+      ],
+    };
+    const sarifFiles = {};
+    const sarifArtifactIndices = {};
+    let nextArtifactIndex = 0;
+    const sarifRules = {};
+    const sarifRuleIndices = {};
+    let nextRuleIndex = 0;
+    const sarifResults = [];
+    const embedFileContents = process.env.SARIF_EMBER_TEMPLATE_LINT_EMBED === 'true';
+
+    const toolConfigurationNotifications = [];
+
+    for (let filePath of Object.keys(results.files)) {
+      let result = results.files[filePath];
+
+      if (typeof sarifFiles[result.filePath] === 'undefined') {
+        sarifArtifactIndices[result.filePath] = nextArtifactIndex++;
+
+        sarifFiles[result.filePath] = {
+          location: {
+            uri: url.pathToFileURL(result.filePath),
+          },
+        };
+
+        if (embedFileContents) {
+          try {
+            // Try to get the file contents and encoding.
+            const contents = fs.readFileSync(result.filePath, { encoding: 'utf-8' });
+
+            sarifFiles[result.filePath].contents = {
+              text: contents,
+            };
+            sarifFiles[result.filePath].encoding = 'utf-8';
+          } catch (error) {
+            this.console.log(error);
+          }
+        }
+
+        if (result.messages.length > 0) {
+          for (const message of result.messages) {
+            const sarifRepresentation = {
+              level: getResultLevel(message),
+              message: {
+                text: message.message,
+              },
+              locations: [
+                {
+                  physicalLocation: {
+                    artifactLocation: {
+                      uri: url.pathToFileURL(result.filePath),
+                      index: sarifArtifactIndices[result.filePath],
+                    },
+                  },
+                },
+              ],
+            };
+
+            if (message.rule && message.rule !== 'global') {
+              sarifRepresentation.ruleId = message.rule;
+
+              if (typeof sarifRules[message.rule] === 'undefined') {
+                sarifRuleIndices[message.rule] = nextRuleIndex++;
+
+                // Create a new entry in the rules dictionary.
+                sarifRules[message.rule] = {
+                  id: message.rule,
+                  helpUri: `https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/${message.rule}.md`,
```

Comment:
Should probably not link to `master`, but instead link to `blog/v${currentVersion}/docs/rule/...`

Reply:
Good call. Done.